src/org/opensolaris/opengrok/web/ProjectHelper.java

```diff
@@ -116,114 +96,354 @@ private void generateTreeMap() throws IOException {
+            val = cfg.isAllowed(group);
+            val = val && !filterGroups(group.getDescendants()).isEmpty();
+            p = (Map<String, Boolean>) cfg.getRequestAttribute("project_helper_allowed_subgroup");
+            p.put(group.getName(), val);
+        }
+        cfg.setRequestAttribute("project_helper_allowed_subgroup", p);
+        return val;
+    }
+
+    /**
+     * Checks if given group contains a favourite project.
+     *
+     * Favourite project is a project which is contained in the OpenGrokProject
+     * cookie, i. e. it has been searched or viewed by the user.
+     *
+     * This should by used to determine if this group should be displayed
+     * expanded or rolled up.
+     *
+     * @param group group
+     * @return true if it has favourite project
+     */
+    @SuppressWarnings(value = "unchecked")
+    public boolean hasFavourite(Group group) {
+        Boolean val;
+        Map<String, Boolean> p = (Map<String, Boolean>) cfg.getRequestAttribute("project_helper_favourite_group");
+        if (p == null) {
+            p = new TreeMap<String, Boolean>();
+            cfg.setRequestAttribute("project_helper_favourite_group", p);
+        }
+        val = p.get(group.getName());
+        if (val == null) {
+            Set<Project> favourite = getAllGrouped();
+            favourite.removeIf(new Predicate<Project>() {
+                @Override
+                public boolean test(Project t) {
+                    // project is favourite
+                    if (!isFavourite(t)) {
+                        return true;
+                    }
+                    // project is contained in group repositories
+                    if (getRepositories(group).contains(t)) {
+                        return false;
+                    }
+                    // project is contained in group projects
+                    if (getProjects(group).contains(t)) {
+                        return false;
+                    }
+                    // project is contained in subgroup's repositories and projects
+                    for (Group g : filterGroups(group.getDescendants())) {
+                        if (getProjects(g).contains(t)) {
+                            return false;
+                        }
+                        if (getRepositories(g).contains(t)) {
+                            return false;
+                        }
+                    }
+                    return true;
+                }
+            });
+            val = !favourite.isEmpty();
+            p.put(group.getName(), val);
+        }
+        cfg.setRequestAttribute("project_helper_favourite_group", p);
+        return val;
+    }
+
+    /**
+     * Checks if the project is a favourite project
+     *
+     * @param project project
+     * @return true if it is favourite
+     */
+    public boolean isFavourite(Project project) {
+        return cfg.getCookieVals("OpenGrokProject").contains(project.getDescription());
```

Comment:
name of the cookie should also be some constant, if it's not, create one in appropriate file please - "OpenGrokProject"